packages/smithy-client/src/date-utils.ts

```diff
@@ -31,5 +34,267 @@ export function dateToUtcString(date: Date): string {
   const minutesString = minutesInt < 10 ? `0${minutesInt}` : `${minutesInt}`;
   const secondsString = secondsInt < 10 ? `0${secondsInt}` : `${secondsInt}`;
 
-  return `${days[dayOfWeek]}, ${dayOfMonthString} ${months[month]} ${year} ${hoursString}:${minutesString}:${secondsString} GMT`;
+  return `${DAYS[dayOfWeek]}, ${dayOfMonthString} ${MONTHS[month]} ${year} ${hoursString}:${minutesString}:${secondsString} GMT`;
 }
+
+const RFC3339 = new RegExp(
+  /^(?<Y>\d{4})-(?<M>\d{2})-(?<D>\d{2})[tT](?<H>\d{2}):(?<m>\d{2}):(?<s>\d{2})(?:\.(?<frac>\d+))?[zZ]$/
+);
+
+/**
+ * Parses a value into a Date. Returns undefined if the input is null or
+ * undefined, throws an error if the input is not a string that can be parsed
+ * as an RFC 3339 date.
+ *
+ * Input strings must conform to RFC3339 section 5.6, and cannot have a UTC
+ * offset. Fractional precision is supported.
+ *
+ * {@see https://xml2rfc.tools.ietf.org/public/rfc/html/rfc3339.html#anchor14}
+ *
+ * @param value the value to parse
+ * @return a Date or undefined
+ */
+export const parseRfc3339DateTime = (value: unknown): Date | undefined => {
+  if (value === null || value === undefined) {
+    return undefined;
+  }
+  if (typeof value !== "string") {
+    throw new TypeError("RFC-3339 date-times must be expressed as strings");
+  }
+  const match = RFC3339.exec(value);
+  if (!match || !match.groups) {
+    throw new TypeError("Invalid RFC-3339 date-time value");
+  }
+  const year = strictParseShort(match.groups["Y"])!;
```

Comment:
nit: You can just use `parseNumber` here to avoid unnecessary validation since you already have asserted that it's exactly 4 digits in your regex. Actually.... now that I think about it you'll probably have to do your own thing that just discards leading zeroes if you want to be strictly correct - so that dates prior to Jan 1st 1000 can be represented.

Reply:
Stripped leading 0s from every component before parsing, and added tests to verify components with leading 0s as well as components that are all 0.